serial: Add UART over Bluetooth LE

ubieda opened the pull request:
### Description
This PR introduces a UART-compatible implementation of Bluetooth LE GATT, using NUS (Nordic UART Service), which allows users to utilize Bluetooth LE as serial endpoints with UART APIs and take advantage of existing support of subsystems that already support UART.

### Background
This PR fixes #12714. Also, this PR is a follow-up of #68076, with the agreed [changes captured on the Bluetooth WG](https://github.com/zephyrproject-rtos/zephyr/pull/68076#issuecomment-1945971326).

### Changes
- Add NUS as a Bluetooth LE GATT Service.
- Add Sample demonstrating API usage of NUS.
- Add UART-compatible driver for Bluetooth LE, using NUS service.
- Add snippet for basic usage of UART over Bluetooth LE.
- Add sample support of UART over BLuetooth LE to `samples/subsys/logging/logger` and `samples/subsys/shell/shell_module`.

### Testing
UART over Bluetooth LE has been tested using [this python-app on the PC-side](https://github.com/ubieda/etales-zephyr/blob/master/scripts/bluetoothcli), and the MCU-side running each of the following samples:

- samples/drivers/uart/echo_bot.
- samples/subsys/console/echo.
- samples/subsys/console/getchar.
- samples/subsys/console/getline.
- samples/subsys/logging/logger.
- samples/subsys/shell/shell_module.

### Note
The feature is marked as experimental as the surface area of what uses UART goes beyond the aforementioned samples, and it would be beneficial to further test it before removing it.

ubieda committed:
bluetooth: services: nus: Add Nordic UART Service

NUS is implemented as a Bluetooth LE service, exchanging data treating
the NUS characteristics as Serial endpoints: RX characteristic to
receive, TX characteristic to send binary packets.

This implementation also enables the ability to define multiple
instances of the NUS Service, analogous to mutliple serial endpoints,
to use each one for different purposes. Unless disabled through Kconfig,
NUS instantiates a default instance, similar to what other services do,
which allows users not interested in using multiple instances, to not
worry about the inherent complexities.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
samples: bluetooth: nus: Add Peripheral NUS sample

Demonstrating basic usage of NUS to exchange data.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
drivers: serial: bt: Add UART over NUS Bluetooth driver

Enables usage Bluetooth LE GATT as a serial endpoint to exchange data
using UART APIs. This implementation is compatible with UART Interrupt
Driven APIs and uses the nus-uart device-tree node properties to
configure FIFO buffers for transmitting and receiving. Defining
multiple instances of the driver is possible and it allows implementing
multiple GATT NUS service instances to exchange data as separate serial
endpoints.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
bluetooth: nus: Add Kconfig option to auto-start Bluetooth LE

Enables compatibility of NUS for codebases that are not
Bluetooth-centric (e.g: Non-bluetooth samples that use UART over
Bluetooth LE for Console, Logging, Shell, others).

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
snippets: Add nus-console snippet

Allowing to be applied on various samples to enable UART over Bluetooth
LE NUS without added complexity.
Tested with nrf52840dk/nrf52840 on the following samples:
- samples/subsys/console/echo.
- samples/subsys/console/getchar.
- samples/subsys/console/getline.
- samples/subsys/logging/logger.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
doc: release: Add information on UART over Bluetooth LE and NUS Service

Add documentation on relevant sections to list addition of Bluetooth LE
service NUS (Nordic UART Service), as well as the UART driver for UART
over Bluetooth LE and snippets to easily enable it.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
samples: logging: Add support for UART over Bluetooth LE NUS

Tested and working with nrf52840dk/nrf52840.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
samples: shell: Add support for UART over Bluetooth LE NUS

Tested and working with nrf52840dk/nrf52840.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

bjarki-trackunit commented on subsys/logging/backends/log_backend_ble.c:
```diff
@@ -1,210 +0,0 @@
-/*
```
I don't agree with removing the log backend for BLE, backends are not dependent on UART by design, they use a way simpler pipe like read/write API, especially compared to injecting a BLE to UART emulator into the chain, so it now looks like BLE -> UART -> shell uart backend, instead of just shell ble backend.

ubieda replied:
Thanks for your feedback, @bjarki-trackunit!

TLDR: I don't have a strong preference on this and we can take it out of the PR if there are disagreements :-).

I agree that is simpler using directly a subsystem backend such as Logging vs abstracting it behind an UART device. I, myself, submitted a PR not so long ago attempting to add Bluetooth LE as a Shell backend (along the same order of ideas). This PR, for various reasons didn't make it upstream (see discussion on #68076).

Taking the feedback from that previous PR: The idea on this one is to go for a common driver that deals directly with the Bluetooth LE GATT Service and interactions, and by doing that, make it easier to maintain the code while serving the same purposes. This approach is inspired by what USB CDC ACM does which implements the device class to act as an UART driver vs implementing on a per-subsystem basis.

On a semi-related data-point: Interestingly enough, the BLE Logging backend was conceived in the context of needing UART over Bluetooth LE, but not implemented due to the author not feeling comfortable dealing with the BLE Stack on such a broad surface. Here's the comment where it originated: https://github.com/zephyrproject-rtos/zephyr/issues/34949#issuecomment-1486449043

In the end, I'm in pro of:
1. Making the feature (Console, Shell, Logging) easily accessible and intuitive to use; which can be achieved by good documentation, consistency, among other things.
2. Making it easier for maintainers to keep code updated and bug-free.

As a result, I consider that maintaining the driver is more sustainable than individual backends.

Nevertheless, if after all the context behind, there are still disagreements on this point, I can take it out of scope and discuss it as a separate issue/PR. Would you agree, @bjarki-trackunit ?

nordicjm commented on snippets/uart_bt/README.rst:
```diff
@@ -0,0 +1,37 @@
+Overview
+********
+
+This snippet redirects serial console and serial shell output to a UART over BLE instance.
+The Bluetooth Serial device used shall be configured using :ref:`devicetree`.
```
This is a duplicate of MCUmgr shell, so the point from the previous discussion stands to discuss why we need a duplicate of this in-tree

jori-nordic replied:
As you said yourself, it's really not a duplicate:
MCUmgr is a binary protocol, with compression, error checking and whole app ecosystem.
NUS is a best-effort data pipe, just like standard UART.

Now obviously MCUmgr is the robust solution here. 
But there is an argument to be made as why we still have serial backends in the zephyr shell, and why "dumb" serial is still used for e.g. the linux boot console: *it's simple*.

There are just less moving pieces and less code overall, so less things that can go wrong or that have to be debugged. This is my argument as to why this "inferior" BLE backend should be allowed in-tree.

nordicjm replied:
If this is using a standard bluetooth UART service then that would be wonderful, please link to the profile if so. If not, then it's irrelevant because a vendor service is a vendor service, which is custom and non-standard.

carlescufi replied:
Bluetooth WG:

First of all, NUS has become over the years a de-facto standard for UART over BLE. This can easily be demonstrated by looking at the sheer number of NUS-compatible client and server implementations in the wild:

https://thejeshgn.com/2016/10/01/uart-over-bluetooth-low-energy/
https://github.com/basilfx/go-ble-nordic-uart
https://www.esp32.com/viewtopic.php?t=2147

Note that there was an attempt to standardize UART over BLE that was unsuccessful, so it is very unlikely, unless Nordic pushes NUS through the spec process, that there will ever be a standard for this.

Furthermore, the overlap between this PR and the functionality provided by mcumgr is minimal: only the actual use of the NUS UART implementation for transporting the console overlaps, the rest is different. Notably, this PR implements the Zephyr UART, which then as a consequence makes it possible to use that UART as a console transport.

I see no issue at all in this particular case to support both transports for the console. They are very different in implementation, the client-side requires different software packages, and perhaps most importantly, users want to be able to use NUS in Zephyr both as a regular UART as well as a console transport.

nordicjm replied:
> First of all, NUS has become over the years a de-facto standard for UART over BLE

Well this is simply not true, it might be used by many projects just because "it is there" but that does not mean anything about standardisation. No different than with e.g. wifi routers use pre-published specs that then are incompatible with the published specification and incompatible with any other device.

Anyhow, requesting for this to be brought up at and discussed at architecture working group

carlescufi replied:
Alright, I will bring it up tomorrow

carlescufi commented on dts/bindings/serial/zephyr,bt-uart.yaml:
```diff
@@ -0,0 +1,19 @@
+
+description: UART over Bluetooth LE
+
+compatible: "zephyr,bt-uart"
```
Bluetooth WG:

Agreement to call this `zephyr,nus-uart`

ubieda replied:
Addressed.

carlescufi commented on snippets/uart_bt/snippet.yml:
```diff
@@ -0,0 +1,4 @@
+name: uart_bt
```
`name: nus-console`
See https://github.com/zephyrproject-rtos/zephyr/blob/main/snippets/cdc-acm-console/cdc-acm-console.conf

ubieda replied:
Addressed.

stephanosio commented:
Re: BabbleSim Tests failure, this has just been fixed in the main. Please rebase to pull in the fix.

carlescufi commented:
Architecture WG:

- @nashif and @henrikbrixandersen state that the duplication here is not significant enough (i.e. it's just a snippet) to justify not including it
- @nashif argues that having a pre-configuration in the form of a snippet could be done by users
- @henrikbrixandersen fails to see how it can be considered harmful to have an extra snippet in the tree

msmttchr commented:
I think this PR add a very useful feature.
I understand it’s adding a proprietary Bluetooth service to handle Uart over BLE derived from nus (Nordic UART service). Being this service proprietary and available to Zephyr users, wouldn’t be more appropriate to call it zus (Zephyr UART service) and explain in the documentation that he has been originally derived from nus?
In my limited experience with zephyr, it seems to me that company specific code is usually found in drivers/soc/boards directories.
Just an observation, feel free to ignore it if it’s not relevant.

ubieda commented:
@msmttchr,

> Being this service proprietary and available to Zephyr users, wouldn’t be more appropriate to call it zus (Zephyr UART service) and explain in the documentation that he has been originally derived from nus?

I do not have a strong preference, other than keeping the UUIDs to ensure compatibility with existing NUS applications, and making this obvious/intuitive for users.

Part of me thinks that renaming it to ZUS makes that association somewhat blurier, but it can be overcome, especially if keeping the UUIDs is an option.

My soft-vote is to keep it NUS, but I'd be curious to hear what others think: do you have thoughts around this, @jori-nordic @carlescufi?

Thanks for your comment!

alwa-nordic commented:
I am also for keeping the name 'NUS'.

ubieda commented:
FYI - Rebased after Non-related CI failures were addressed on #70215. CI Checks now passing.

jhedberg commented on drivers/serial/Kconfig.bt:
```diff
@@ -0,0 +1,29 @@
+# Disable default NUS instance to not confuse it with the instances defined on UART_BT
+config BT_NUS_DEFAULT_INSTANCE
+	bool
+	default n
```
minor: n is the "default default" so this line is redundant

ubieda replied:
This only overrides the default value of `BT_NUS_DEFAULT_INSTANCE` to `n` when `UART_BT` is enabled. The actual Kconfig is defined under `subsys/bluetooth/services/nus/Kconfig`, and its purpose is to allow users to use NUS as a standalone GATT Service.

This means:
- When `BT_NUS` is enabled, but `UART_BT` is not, the service is instantiated within the `BT_NUS` module.
- When both `BT_NUS` and `UART_BT `are enabled, we don't have a default instance and we expect the user to define `nus-uart` DTS instances.

I'll add a more descriptive comment on top of this Kconfig symbol to reflect this purpose, but LMK if there's a better way to achieve this goal.

jhedberg replied:
Oh, I see. I haven't seen this kind of usage of Kconfig overlays elsewhere. Normally (AFAIK) those are used in board files and application overlays, but not in drivers or subsystems. Would it perhaps make sense to make the original definition conditional to `UART_BT`, i.e. instead of `default y` you'd have `default y if !UART_BT` there?

ubieda replied:
That sounds good. I'll go ahead and make that change, instead.

jhedberg commented on drivers/serial/uart_bt.c:
```diff
@@ -0,0 +1,320 @@
+
+static int uart_bt_init(const struct device *dev)
+{
+	int err = -1;
```
Unnecessary initialization of the variable here (it's always unconditionally set in the function before trying to read it)

jhedberg commented on drivers/serial/uart_bt_auto_start_bt.c:
```diff
@@ -0,0 +1,43 @@
+	ARG_UNUSED(b);
+	ARG_UNUSED(c);
+
+	int err = -1;
```
Unnecessary initialization here as well.

jhedberg commented on samples/bluetooth/peripheral_nus/src/main.c:
```diff
@@ -0,0 +1,82 @@
+
+int main(void)
+{
+	int err = -1;
```
And here. Btw, even if it was justitified to initialize it, using `-1` is not really a good idea. Zephyr generally uses negative posix errno values for `int` type errors, so there's a risk someone might interpret `-1` as `-EPERM`.

jhedberg reviewed (changes requested):
A few minor comments about error values, but generally this looks pretty good to me.

ubieda commented:
I've made the following changes:
- Removed unnecessary initializations for `err`.
- `BT_NUS_DEFAULT_INSTANCE` is no longer overriden on Kconfig.bt, as suggested.
- (I added this one even though hasn't been requested) Added ability to specify connection object through NUS TX/RX APIs. It allows users to interact with a single peer through if NUS is used standalone. There's no functional change, but rather allows more future flexibility without API changes.

msmttchr commented:
> @msmttchr,
> 
> > Being this service proprietary and available to Zephyr users, wouldn’t be more appropriate to call it zus (Zephyr UART service) and explain in the documentation that he has been originally derived from nus?
> 
> I do not have a strong preference, other than keeping the UUIDs to ensure compatibility with existing NUS applications, and making this obvious/intuitive for users.
> 
> Part of me thinks that renaming it to ZUS makes that association somewhat blurier, but it can be overcome, especially if keeping the UUIDs is an option.
> 
> My soft-vote is to keep it NUS, but I'd be curious to hear what others think: do you have thoughts around this, @jori-nordic @carlescufi?
> 
> Thanks for your comment!

Ok, thanks for the explanation of the rationale. Clear to me.

jori-nordic commented on drivers/serial/uart_bt_auto_start_bt.c:
```diff
@@ -0,0 +1,43 @@
+/*
```
This has technically nothing to do with uart, it's pretty much like a systemd service file for BLE.
But it's hard to find a place where it would belong, yeah.

Maybe in ~/repos/zephyrproject/zephyr/subsys/bluetooth/lib` ? or next to the nus-service.

jori-nordic commented on drivers/serial/uart_bt_auto_start_bt.c:
```diff
@@ -0,0 +1,43 @@
+	__ASSERT_NO_MSG(!err);
+}
+
+K_THREAD_DEFINE(uart_bt_auto_start_id,
```
Can sys_init with the `APPLICATION` prio be used here instead to save a one-shot thread?
I'm not sure, just asking if you can try and see if it's even possible

jori-nordic reviewed (commented):
Haven't tested it yet, will do tomorrow and then approve if it works.
I don't have other comments apart from those (not NAKs).

ubieda commented on include/zephyr/bluetooth/services/nus.h:
```diff
@@ -0,0 +1,135 @@
+
+/** @brief Send Data over NUS
+ *
+ * @note This API sends the data to all connected peers.
```
Whoops - Outdated comment. Sorry. I'll update it

ubieda commented on samples/bluetooth/peripheral_nus/src/main.c:
```diff
@@ -0,0 +1,83 @@
+
+	err = bt_nus_cb_register(&nus_listener, NULL);
+	if (err) {
+		return err;
```
Will do.

ubieda commented on dts/bindings/serial/zephyr,nus-uart.yaml:
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2024 Croxel, Inc.
```
I can lump this commit together with the addition of UART over NUS.

ubieda commented:
> Haven't tested it yet, will do tomorrow and then approve if it works. I don't have other comments apart from those (not NAKs).

@jori-nordic Thanks for reviewing. I'll address your comments/suggestions today.

ubieda commented:
I've made the following changes:
- Update outdated comments on `bt_nus_send()` and `bt_nus_inst_send()` APIs.
- Add printks on errors on `peripheral_nus` sample.
- Moved Bluetooth auto-init code to `subsys/bluetooth/services/nus`.
- Changed Bluetooth auto-init code to execute on `SYS_INIT()`.
- Updated release notes to match updated Kconfig symbol for `CONFIG_BT_NUS_AUTO_START_BLUETOOTH`.

jori-nordic commented:
Logging + the echo (ie console) works w/ your script and the nrf connect apps.
The shell however doesn't. I tried with both nrf apps, and the script. I compiled `shell_module` both ways and see no output, nothing is even echoed back.
```bash
# using the snippet
west build -b nrf52840dk/nrf52840 -S nus-console
# using the overlays
west build -b nrf52840dk/nrf52840 -- -DDTC_OVERLAY_FILE=bt.overlay -DOVERLAY_CONFIG=overlay-bt.conf
```

alwa-nordic commented on drivers/serial/uart_bt.c:
```diff
@@ -0,0 +1,321 @@
+		 * size per-connection. This has the trade-off of limiting
+		 * throughput but allows multi-connection support.
+		 */
+		len = ring_buf_get_claim(dev_data->uart.tx_ringbuf, &data, 23);
```
We must subtract the ATT_HANDLE_VALUE_NTF overhead (3 bytes) from the minimum ATT_MTU (23 bytes) to get the useful MTU.

```suggestion
		len = ring_buf_get_claim(dev_data->uart.tx_ringbuf, &data, 20);
```

ubieda replied:
Good catch - Thanks for pointing it out. I can address it on a follow-up PR if that's ok!

ubieda replied:
Addressed by #70668. Resolving now.

alwa-nordic commented on drivers/serial/uart_bt.c:
```diff
@@ -0,0 +1,321 @@
+		LOG_ERR("RX Ring buffer full. received: %d, added to queue: %d", len, put_len);
+	}
+
+	k_work_submit(&dev_data->uart.cb_work);
```
Should this be guarded by `if (dev_data->uart.rx_irq_ena)`?

alwa-nordic replied:
This can also optionally be followed up in a separate PR. It is unclear if spurious interrupts are allowed in the UART IRQ API. If you do take a look, you may also want to guard against TX interrupts from `uart_bt_irq_tx_enable` when the buffer is full. Maybe it's over-engineering.. but correctness! :)

ubieda replied:
I'll take a closer look at this when following up on suggested improvements (separate PR). This is good feedback. Thanks!

alwa-nordic commented:
Just wanted to say I think this high quality code. :+1: The code comments are on point - it makes it easy to read. And the documentation is good as well.
